### PR TITLE
Rewrite Docs Generator from Python to Scala/Java

### LINKED ISSUE #1699: Doc Chunk Generator : Traverse through std-lib to generate `docs` folder
### Summary

Generally port code from `reference` repository from Python to Java/Scala

### PR DESCRIPTION
**Pull request by BinarySoftware**
### Pull Request Description
Closes #1699 

It generates HTML files for future use in IDE, as well as Js files with tree of contents and react elements for use on website.

Also, I've refactored a bit Doc Generator in syntax-specialization library. Just made it cleaner, no changes to functionality.
### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [ ] All code has been tested where possible.

### CODE REVIEW

File: lib/scala/docs-generator/src/main/scala/org/enso/DocsGeneratorMain.scala [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import scala.io.Source
+import scalatags.Text.{all => HTML}
+import HTML._
+import docsgenerator.DocsGenerator._
+import docsgenerator.TreeOfCommonPrefixes._
+
+/** The Docs Generator script.
+  * Used to create HTML documentation for Enso Standard Library, and also JS
+  * files with react components for the "component reference" website.
+  */
```
Comment (iamrecursion) [inline lib/scala/docs-generator/src/main/scala/org/enso/DocsGeneratorMain.scala]:
```suggestion
/** The entry point for the documentation generator.
  *
  * The documentation generator is responsible for creating HTML documentation for the Enso standard library. It also generates JavaScript files containing react components for the reference website.
  */
```
The text "reference website" should link to https://enso.org/docs/reference.

File: lib/scala/docs-generator/src/main/scala/org/enso/DocsGeneratorMain.scala [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+object DocsGeneratorMain extends App {
+
+  /// Files
+  val path = "./distribution/std-lib/Standard/src"
```
Comment (iamrecursion) [inline lib/scala/docs-generator/src/main/scala/org/enso/DocsGeneratorMain.scala]: Nope. You're doing the same thing we discussed with the python version and hardcoding a bunch of constants. Things like this should be accepted via the command-line (using our CLI args framework), not hard-coded in the source.

File: lib/scala/docs-generator/src/main/scala/org/enso/DocsGeneratorMain.scala [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    .toList
+  val allDocs = allPrograms
+    .map(s => generate(s.getOrElse("")))
+    .map(mapIfEmpty)
+    .map(removeUnnecessaryDivs)
+  val allDocFiles = allFiles.map(
+    _.getPath
+      .replace(".enso", ".html")
+      .replace("Standard/src", "docs")
+  )
```
Comment (iamrecursion) [inline lib/scala/docs-generator/src/main/scala/org/enso/DocsGeneratorMain.scala]: Do keep in mind that we don't actually want this rendered HTML. From the engine perspective we care about the chunks so we can assemble actually useful blocks of documentation that include extension methods, re-exports, and so on.

File: lib/scala/docs-generator/src/main/scala/org/enso/docsgenerator/DocsGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import org.enso.syntax.text.{DocParser, Parser}
+import org.enso.syntax.text.docparser._
+
+/** The Docs Generator class. Defines useful wrappers for Doc Parser.
+  */
+object DocsGenerator {
+
+  /** Generates HTML of docs from Enso program.
+    */
+  def generate(program: String): String = {
```
Comment (iamrecursion) [inline lib/scala/docs-generator/src/main/scala/org/enso/docsgenerator/DocsGenerator.scala]: Call this `run` given the object is `DocsGenerator`. `DocsGenerator.run` is much nicer than `DocsGenerator.generate`.

File: lib/scala/docs-generator/src/main/scala/org/enso/docsgenerator/TreeOfCommonPrefixes.scala [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  */
+object TreeOfCommonPrefixes {
+
+  /** A single tree node of given `name` containing other nodes as list called
+    * `elems`.
+    */
+  case class Node(name: String, var elems: List[Node]) {
+
+    /** Generates HTML tree from node, as this method is used to create page
+      * chooser.
```
Comment (iamrecursion) [inline lib/scala/docs-generator/src/main/scala/org/enso/docsgenerator/TreeOfCommonPrefixes.scala]:
```suggestion
    /** Generates an HTML tree from node, as this method is used to create the page
      * chooser.
```

File: lib/scala/docs-generator/src/main/scala/org/enso/docsgenerator/TreeOfCommonPrefixes.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+          HTML.label(HTML.`for` := newBeg)(
+            HTML.a(HTML.href := newBeg)(name)
+          ),
+          elems.map(x => HTML.ul(x.html(newBeg)))
+        )
+      }
+    }
+  }
+
+  /** Groups a list of nodes by it's prefixes. Calls groupByPrefix on every node.
```
Comment (iamrecursion) [inline lib/scala/docs-generator/src/main/scala/org/enso/docsgenerator/TreeOfCommonPrefixes.scala]:
```suggestion
  /** Groups a list of nodes by it's prefixes.
```

File: lib/scala/docs-generator/src/main/scala/org/enso/docsgenerator/TreeOfCommonPrefixes.scala [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    */
+  def groupNodesByPrefix(le: List[Node]): List[Node] =
+    groupByPrefix(le.map(_.name))
+
+  /** Recursive function that groups a list of strings by it's prefixes, actually
+    * creating the aforementioned tree.
+    */
```
Comment (iamrecursion) [inline lib/scala/docs-generator/src/main/scala/org/enso/docsgenerator/TreeOfCommonPrefixes.scala]: You don't need to say that it's a recursive function. Documentation is about the "why", not the "what".

File: lib/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/docparser/DocParserRunner.scala [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+import org.enso.data.List1
+import org.enso.syntax.text.{AST, DocParser}
+import org.enso.syntax.text.Shape.Block.Line
+import org.enso.syntax.text.ast.Doc
+
+////////////////////////////////////////////////////////////////////////////////
+//// Doc Parser Runner /////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+
+/** This is Doc Parser Runner.
```
Comment (iamrecursion) [inline lib/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/docparser/DocParserRunner.scala]: This is useless.

File: lib/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/docparser/DocParserRunner.scala [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+  *
+  * Essentially it binds together Enso Parser with Doc Parser.
+  * When Parser finishes its job it invokes runner with AST created by it after
+  * resolving macros. Then Runner does it's job - running Doc Parser on every
+  * [[AST.Comment]], combined with connecting [[Doc]] with AST in
+  * [[AST.Documented]] node, which gets AST from [[AST.Def]] and
+  * [[AST.App.Infix]]
+  */
+object DocParserRunner {
+  //////////////////////////////////////////////////////////////////////////////
```
Comment (iamrecursion) [inline lib/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/docparser/DocParserRunner.scala]:
```suggestion
object DocParserRunner {

  //////////////////////////////////////////////////////////////////////////////
```

File: lib/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/docparser/DocParserRunner.scala [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+  //// Created Doc's in right places with appropriate AST //////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+  /** This function invokes the documentation parser on every instance of
```
Comment (iamrecursion) [inline lib/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/docparser/DocParserRunner.scala]:
```suggestion
  //////////////////////////////////////////////////////////////////////////////
  
  /** This function invokes the documentation parser on every instance of
```

**Review by iamrecursion — CHANGES_REQUESTED** [review]
Somebody else needs to review the JavaScript and CSS files.

File: lib/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/docparser/DocParserRunner.scala (hunk 9 above)
Comment (BinarySoftware) [reply lib/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/docparser/DocParserRunner.scala]: Can't add 2 empty lines before section, scalafmt automatically reverts into 1 empty line between.

**Review by iamrecursion — CHANGES_REQUESTED** [review]

File: lib/scala/docs-generator/src/main/scala/org/enso/docs/generator/Main.scala [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+  * [[https://enso.org/docs/reference reference website]].
+  */
+object Main {
+
+  var libPath =
+    "./lib/scala/docs-generator/src/main/scala/org/enso/docs/generator/"
+  var path           = "./distribution/std-lib/Standard/src"
+  var jsTempFileName = "template.js"
+  var cssFileName    = "treeStyle.css"
+  var outDir         = "docs-js"
```
[the remark this answers is not in the document]
Comment (BinarySoftware) [reply lib/scala/docs-generator/src/main/scala/org/enso/docs/generator/Main.scala]: It needs to know where the templates are located. renamed it so it's more obvious what it is.

**Review by iamrecursion — APPROVED** [review]

**Pull request merged**